File: JavaScript/bogoSort.js
/**
 * Function to check if the array is sorted in ascending order.
 * @param {number[]} arr - The array to check.
 * @returns {boolean} - True if the array is sorted, false otherwise.
 */
function isSorted(arr) {
    for (let i = 1; i < arr.length; i++) {
        if (arr[i] < arr[i - 1]) {
            return false; // If any element is less than the previous one, return false
        }
    }
    return true; // Array is sorted
}

/**
 * Function to shuffle the array randomly.
 * @param {number[]} arr - The array to shuffle.
 * @returns {number[]} - The shuffled array.
 */
function shuffle(arr) {
    for (let i = arr.length - 1; i > 0; i--) {
        // Generate a random index
        const j = Math.floor(Math.random() * (i + 1));
        // Swap the elements at indices i and j
        [arr[i], arr[j]] = [arr[j], arr[i]];
    }
    return arr;
}

/**
 * Function to perform Bogo Sort on an array.
 * @param {number[]} arr - The array to be sorted.
 * @returns {number[]} - The sorted array.
 */
function bogoSort(arr) {
    // Repeat until the array is sorted
    while (!isSorted(arr)) {
        shuffle(arr); // Shuffle the array randomly
    }
    return arr; // Return the sorted array
}

// Example usage
const arrayToSort = [3, 1, 4, 2, 5];
const sortedArray = bogoSort(arrayToSort);
console.log("Sorted Array:", sortedArray);

File: Go/bogoSort.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

// isSorted checks if the array is sorted in ascending order.
func isSorted(arr []int) bool {
	for i := 1; i < len(arr); i++ {
		if arr[i-1] > arr[i] {
			return false // Return false if any element is greater than the next
		}
	}
	return true // Return true if the array is sorted
}

// shuffle randomly shuffles the elements of the array.
func shuffle(arr []int) {
	rand.Seed(time.Now().UnixNano()) // Seed the random number generator
	for i := range arr {
		j := rand.Intn(len(arr)) // Generate a random index
		arr[i], arr[j] = arr[j], arr[i] // Swap elements at indices i and j
	}
}

// bogoSort repeatedly shuffles the array until it is sorted.
func bogoSort(arr []int) {
	for !isSorted(arr) { // Continue until the array is sorted
		shuffle(arr) // Shuffle the array
	}
}

// main function to test the bogoSort
func main() {
	arr := []int{3, 2, 5, 1, 4} // Sample array to be sorted
	fmt.Println("Original array:", arr)

	bogoSort(arr) // Call the sorting function

	fmt.Println("Sorted array:", arr) // Print the sorted array
}
